vyper/builtin_functions/functions.py

```diff
@@ -2520,6 +2526,109 @@ def build_IR(self, expr, args, kwargs, context):
             )
 
 
+class Append(BuiltinFunction):
+    _id = "append"
+
+    def build_IR(self, expr, context):
+        darray = Expr(expr.func.value, context).ir_node
+        args = [Expr(x, context).ir_node for x in expr.args]
+
+        # sanity checks
+        assert len(args) == 1
+        arg = args[0]
+        assert isinstance(darray.typ, DArrayType)
+
+        check_assign(dummy_node_for_type(darray.typ.subtype), dummy_node_for_type(arg.typ))
+
+        assert darray.typ.count > 0, "jerk boy u r out"
+
+        ret = ["seq"]
+        with darray.cache_when_complex("darray") as (b1, darray):
+            len_ = get_dyn_array_count(darray)
+            with len_.cache_when_complex("old_darray_len") as (b2, len_):
+                ret.append(["assert", ["lt", len_, darray.typ.count]])
+                ret.append(STORE(darray, ["add", len_, 1]))
+                # NOTE: typechecks elem_node
+                # NOTE skip array bounds check bc we already asserted len two lines up
+                ret.append(
+                    make_setter(get_element_ptr(darray, len_, array_bounds_check=False), arg)
+                )
+                return IRnode.from_list(b1.resolve(b2.resolve(ret)))
+
+
+class Extend(BuiltinFunction):
+    _id = "extend"
+
+    def build_IR(self, expr, context):
+        dst = Expr(expr.func.value, context).ir_node
+        # sanity checks
+        assert len(expr.args) == 1
+        src = Expr(expr.args[0], context).ir_node
+
+        assert isinstance(dst.typ, DArrayType)
+        assert isinstance(src.typ, DArrayType)
+
+        ret = ["seq"]
+
+        with dst.cache_when_complex("darray_dst") as (b1, dst):
+            dst_len = get_dyn_array_count(dst)
+
+            with dst_len.cache_when_complex("dst_darray_len") as (b2, dst_len):
+
+                dst_bound = dst.typ.count
+                src_len = get_dyn_array_count(src)
+                new_len = IRnode.from_list(["add", dst_len, src_len], typ="uint256")
+
+                # Assert that `src_len + dst_len` <= maxlen(dst)`
+                check = IRnode.from_list(["assert", ["le", new_len, dst_bound]])
+                ret.append(check)
+
+                # Store updated length
+                store_length = IRnode.from_list(STORE(dst, new_len))
+                ret.append(store_length)
+
+                # Get start pointer of dst
+                dst_start_idx = get_element_ptr(dst, dst_len, array_bounds_check=False)
+
+                # Cast dst start pointer as darray for `_dynarray_make_setter` by subtracting offset
+                dst_i = IRnode.from_list(["sub", dst_start_idx, dst.location.word_scale])
```

Comment:
kind of bothers me that we still have to adjust the offset like this for `copy_dynarray_body`